File: game/data/plugins/skill/herblore/herb.rb
require 'java'

java_import 'org.apollo.game.action.Action'
java_import 'org.apollo.util.LanguageUtil'

# A herb is an ingredient that requires identification before being used.
class Herb < Ingredient
  include HerbloreMethod

  attr_reader :unidentified, :level, :experience

  def initialize(item_id, unidentified, level, experience)
    super item_id

    @unidentified = unidentified
    @level = level
    @experience = experience
  end

  def invoke(player, _id, slot)
    item = player.inventory.get(slot)
    player.start_action(HerbIdentificationAction.new(player, self, slot, item))
  end
end

# An action that makes a player identify a herb.
class HerbIdentificationAction < Action
  attr_reader :herb, :slot, :item, :pulses

  def initialize(player, herb, slot, item)
    super(0, true, player)

    @herb = herb
    @slot = slot
    @item = item
    @pulses = 0
  end

  def execute
    if @pulses == 0
      unless check_skill(mob, @herb.level, 'identify this herb')
        stop
        return
      end
    end

    execute_action
    @pulses += 1
  end

  def execute_action
    player = mob
    inventory = player.inventory

    if inventory.remove_slot(@slot, 1) == 1
      identified = @herb.item

      inventory.add(identified)

      article = LanguageUtil.getIndefiniteArticle(identified.definition.name)
      player.skill_set.add_experience(Skill::HERBLORE, @herb.experience)
      player.send_message("This herb is #{article} #{identified.definition.name}.")
    end

    stop
  end

  def equals(other)
    get_class == other.get_class && slot == other.slot && herb == other.herb
  end
end

# Appends a herb to the InventoryItemMessage interception.
def append_herb(item_id, unidentified, level, experience)
  herb = Herb.new(item_id, unidentified, level, experience)
  append_herblore_item(herb, unidentified)
  herb
end

# Herbs

GUAM_LEAF   = append_herb(249,  199,  1,  2.5)
MARRENTILL  = append_herb(251,  201,  5,  3.8)
TARROMIN    = append_herb(253,  203,  11, 5)
HARRALANDER = append_herb(255,  205,  20, 6.3)
RANARR      = append_herb(257,  207,  25, 7.5)
TOADFLAX    = append_herb(2998, 3049, 30, 8)
IRIT_LEAF   = append_herb(259,  209,  40, 8.8)
AVANTOE     = append_herb(261,  211,  48, 10)
KWUARM      = append_herb(263,  213,  54, 11.3)
SNAPDRAGON  = append_herb(3000, 3051, 59, 11.8)
CADANTINE   = append_herb(265,  215,  65, 12.5)
LANTADYME   = append_herb(2481, 2485, 67, 13.1)
DWARF_WEED  = append_herb(267,  217,  70, 13.8)
TORSTOL     = append_herb(269,  219,  75, 15)

File: game/plugin/skills/herblore/src/Herb.kt
import org.apollo.game.plugin.api.Definitions

enum class Herb(
    val identified: Int,
    val unidentified: Int,
    val level: Int,
    val experience: Double
) {
    GUAM_LEAF(identified = 249, unidentified = 199, level = 1, experience = 2.5),
    MARRENTILL(identified = 251, unidentified = 201, level = 5, experience = 3.8),
    TARROMIN(identified = 253, unidentified = 203, level = 11, experience = 5.0),
    HARRALANDER(identified = 255, unidentified = 205, level = 20, experience = 6.3),
    RANARR(identified = 257, unidentified = 207, level = 25, experience = 7.5),
    TOADFLAX(identified = 2998, unidentified = 2998, level = 30, experience = 8.0),
    IRIT_LEAF(identified = 259, unidentified = 209, level = 40, experience = 8.8),
    AVANTOE(identified = 261, unidentified = 211, level = 48, experience = 10.0),
    KWUARM(identified = 263, unidentified = 213, level = 54, experience = 11.3),
    SNAPDRAGON(identified = 3000, unidentified = 3051, level = 59, experience = 11.8),
    CADANTINE(identified = 265, unidentified = 215, level = 65, experience = 12.5),
    LANTADYME(identified = 2481, unidentified = 2485, level = 67, experience = 13.1),
    DWARF_WEED(identified = 267, unidentified = 217, level = 70, experience = 13.8),
    TORSTOL(identified = 269, unidentified = 219, level = 75, experience = 15.0);

    val identifiedName by lazy { Definitions.item(identified)!!.name }

    companion object {
        private val identified = Herb.values().map(Herb::identified).toHashSet()
        private val herbs = Herb.values().associateBy(Herb::unidentified)

        operator fun get(id: Int): Herb? = herbs[id]
        internal fun Int.isUnidentified(): Boolean = this in herbs
        internal fun Int.isIdentified(): Boolean = this in identified
    }
}